### PR TITLE
Add support for HMAC based on sha256 (128b keys).

### PR DESCRIPTION
**Pull request by kjopek**
Add support for HMAC based on SHA256 with 128 bits keys.

It can be used in the following form:
./tcpdump -nr ~/ipsec_dump.pcap -s0 -E "0xXXXXXXXX@A.B.C.D aes256-cbc-hmac-sha256-128:[KEY1],0xXXXXXXXX@E.F.G.H aes256-cbc-hmac-sha256-128:[KEY2]"

Sponsored by: Digital Fingerprint

### CODE REVIEW

File: print-esp.c [hunk 1]
```diff
@@ -332,6 +332,20 @@ int espprint_decode_hex(netdissect_options *ndo,
 	return i;
 }
 
+static int
+strendswith(char *str, char *end)
+{
+	char *p;
+
+	if (strlen(str) > strlen(end) &&
+	    !strcmp(str + strlen(str) - strlen(end), end)) {
```
Comment (mcr) [inline print-esp.c:341]: perhaps a comment or two here explaining the logic.   A temp variable or two for strlen(XX) would also increase readability.

File: print-esp.c [hunk 2]
```diff
@@ -353,18 +367,13 @@ espprint_decode_encalgo(netdissect_options *ndo,
 	}
 	*colon = '\0';
 
-	if (strlen(decode) > strlen("-hmac96") &&
-	    !strcmp(decode + strlen(decode) - strlen("-hmac96"),
-		    "-hmac96")) {
-		p = strstr(decode, "-hmac96");
-		*p = '\0';
+	if (strendswith(decode, "-hmac96"))
```
Comment (mcr) [inline print-esp.c:370]: I see, you inherited this convoluted strlen(). Did I write that 20 years ago?  Maybe.  Thanks for improving upon this.

**Review by mcr — CHANGES_REQUESTED** [review]
Good changes, but it needs some test cases to exercise sha256.
Even if you haven't got any ESP packets, at least, you can instantiate the hash and it can fail to authenticate the packets in question.

**Comment by fxlb** [comment]
And please, rebase on top of the-tcpdump-group:master.

File: print-esp.c (hunk 1 above)
Comment (kjopek) [reply print-esp.c]: Sure, thanks!

**Comment by infrastation** [comment]
In addition to the previous comment, which still stands (merging from 3 years ago is not rebasing), you might want to read the file `CONTRIBUTING.md` (especially step 4), so you have a compiler feedback loop in your working copy.

**Comment by kjopek** [comment]
Thank you very much for all your comments and I am very sorry for breaking contributors' rules.

From what I see now, after 3+ years, this patch was wrong from the very beginning and here is my explanation why I think so.

The original patch was kind-of quick and therefore dirty hack to allow IPSEC debugging under certain configuration. Particularly, to make sure the encryption works as expected and data is correctly encrypted/decrypted. In my humble opinion, the pull request should have been splitted into at least two smaller ones:

1. First one to introduce `strendswith` function and use it where needed.
2. Second one to introduce additional primitive for packet authentication.

While the first one is rather obvious, the second one is tricky as the whole function relies on `authlen` variable which is used only for keeping length of hmac code and is required only to calculate correct offsets in the packet to **decrypt** it correctly. I see no other usage of authentication-related code in `print-esp.c` file. `sa_list` structure maintains `authsecret` and `authsecret_len` fields which are unpopulated and therefore does not play any role in printing ESP packets. In my humble opinion, tcpdump lacks hmac verification and this functionality could be added separately.

Summing up, my original intention was to enable ESP decryption for different ENC+AUTH scenario, but from my current perspective this pull request should be closed and I should create another one only to implement `strendswith` function. The remaining part (lack of auth support) should be discussed and added separately. I am leaving this PR open to allow further discussion.

**Comment by infrastation** [comment]
Thank you for the detailed comment. If you think it would make more sense to split the change into two commits, split it. If the first commit is useful on its own, it can be merged on its own. Would you like to take more time to state the decryption problem properly?

**Comment by ZsBT** [comment]
+1 for this (and keeping the issue open)
